README.md
*This is fork of [archived repository](https://github.com/autoborna/Transifex-API) by [Michael Babker](https://github.com/mbabker). Autoborna use this library and so we decided to maintain it.*

Transifex API Package [![Scrutinizer Code Quality](https://scrutinizer-ci.com/g/autoborna/Transifex-API/badges/quality-score.png?b=master)](https://scrutinizer-ci.com/g/autoborna/Transifex-API/?branch=master) [![Code Coverage](https://scrutinizer-ci.com/g/autoborna/Transifex-API/badges/coverage.png?b=master)](https://scrutinizer-ci.com/g/autoborna/Transifex-API/?branch=master) [![Build Status](https://scrutinizer-ci.com/g/autoborna/Transifex-API/badges/build.png?b=master)](https://scrutinizer-ci.com/g/autoborna/Transifex-API/build-status/master) [![Build Status](https://travis-ci.org/autoborna/Transifex-API.svg?branch=master)](https://travis-ci.org/autoborna/Transifex-API)
===============

[![Latest Stable Version](https://poser.pugx.org/autoborna/transifex/v/stable.svg)](https://packagist.org/packages/autoborna/transifex)
[![Total Downloads](https://poser.pugx.org/autoborna/transifex/downloads.svg)](https://packagist.org/packages/autoborna/transifex)
[![Latest Unstable Version](https://poser.pugx.org/autoborna/transifex/v/unstable.svg)](https://packagist.org/packages/autoborna/transifex)
[![License](https://poser.pugx.org/autoborna/transifex/license.svg)](https://packagist.org/packages/autoborna/transifex)

The *Transifex API Package* is a PHP client for accessing the [Transifex API](http://docs.transifex.com/api/).

Requirements
------------

* PHP 7.4 or later
* Any [PSR-17 compatible](https://www.php-fig.org/psr/psr-17/) factories
* Any [PSR-18 compatible](https://www.php-fig.org/psr/psr-18/) HTTP client

Installation
------------

## Installation via GIT

Get the source code from GIT:

```sh
git clone git://github.com/autoborna/Transifex-API.git
```

## Installation via Composer

To include this package in your Composer project, run the following command from the command line:

```sh
composer require autoborna/transifex
```

Documentation
-------------

The Transifex API package provides a PHP interface for interacting with the [Transifex API](https://developers.transifex.com/reference/api-introduction).

### Basic Use

The primary interface for interacting with the Transifex package is the `Transifex` class. This class serves as the API factory and allows developers to manage the options used by the API objects and HTTP connector as well as retrieve instances of the API objects. To create a `Transifex` object, you only need to instantiate it with the appropriate dependencies.

```php
use Autoborna\Transifex\ApiFactory;
use Autoborna\Transifex\Config;
use Autoborna\Transifex\Transifex;
use Psr\Http\Client\ClientInterface;
use Psr\Http\Message\RequestFactoryInterface;
use Psr\Http\Message\StreamFactoryInterface;
use Psr\Http\Message\UriFactoryInterface;

$client = new ClientInterface(); // or any PSR-18 HTTP client
$requestFactory = new RequestFactoryInterface(); // or any PSR-17 Request factory
$streamFactory = new StreamFactoryInterface(); // or any PSR-17 Stream factory
$uriFactory = new UriFactoryInterface(); // or any PSR-17 URI factory
$config = new Config();
$config->setApiToken('some-api-token');
$config->setOrganization('some-organization');
$config->setProject('some-project');
$transifex = new Transifex($client, $requestFactory, $streamFactory, $uriFactory, $config);
```

[Follow the Transifex documentation on how to get the API Token.](https://developers.transifex.com/reference/api-authentication)

The `Config` object can be also created from environmental variables:
```php
use Autoborna\Transifex\Config;
putenv('TRANSIFEX_API_TOKEN=some-api-token');
putenv('TRANSIFEX_ORGANIZATION=some-organization');
putenv('TRANSIFEX_PROJECT=some-project');
$config = Config::fromEnv();
```

To retrieve an instance of an API object, you would use the `get()` method. API objects are named based on the documented sections of the Transifex API. To retrieve an object that can interface with the "resources" API section, you would use the following code:

```php
use \Autoborna\Transifex\Connector\Resources;
$resources = $transifex->getConnector(Resources::class);
\assert($resources instanceof Resources);
```

### API Responses

This package returns a [PSR-7 compatible](https://www.php-fig.org/psr/psr-7/) response created by the underlying [PSR-18 HTTP client](https://www.php-fig.org/psr/psr-18/).

This package is not catching Exceptions thrown by the HTTP client so users implementing this package should implement appropriate error handling mechanisms.

An exception of `Autoborna\Transifex\Exception\ResponseException` will be thrown when the response code is > 400. The error message contains the URI, request and response. This simplifies debugging. The Request and Response object can be retreived from this exception.

### Resources

Resources are the basic files that should be translated.

Get the Resource connector for the bellow examples like so:
```php
use Autoborna\Transifex\Connector\Resources;

$resources = $transifex->getConnector(Resources::class);
\assert($resources instanceof Resources);
```

#### Get all resources

Docs: [https://developers.transifex.com/reference/get_resources](https://developers.transifex.com/reference/get_resources)

```php
$response = $resources->getAll();
$body     = json_decode($response->getBody()->__toString(), true);
```

#### Create a resource

This is a syncronous operation so you get the response rightaway. You need another request to upload a content to this resource though. See the [docs](https://developers.transifex.com/reference/post_resources) for example responses.

For the list of available i18n formats (the third param) trigger [this endpoint](https://developers.transifex.com/reference/get_i18n-formats).

```php
$response = $resources->create('Resource A', 'resource-a', 'INI');
$body     = json_decode($response->getBody()->__toString(), true);
```

#### Upload resource content

This is an asynchronous operation so the first request will just add the task to the queue and returns a unique ID See the [docs](https://developers.transifex.com/reference/post_resource-strings-async-uploads). Then we have to poll another endpoint for the status of this task.

```php
use Autoborna\Transifex\Promise;
use Autoborna\Transifex\Exception\ResponseException;
use Psr\Http\Message\ResponseInterface;

$response = $resources->uploadContent('resource-a', "something=\"Something\"\nsomething.else=\"Something Else\"\n");
$promise  = $transifex->getApiConnector()->createPromise($response);
$promises = new \SplQueue();
$promise->setFilePath('/some/file.ini'); // In the real world it is useful to map your file to this promise. Useful for later processing.
$promises->enqueue($promise); // In the real world, there would be multiple promises.
usleep(500000); // Give Transifex a 1/2 second so we make 1 request instead of 2.
$transifex->getApiConnector()->fulfillPromises(
    $promises,
    function (ResponseInterface $response, Promise $promise) {
        echo "Resource for {$promise->getFilePath()} was uploaded successfully";
    },
    function (ResponseException $exception, Promise $promise) {
        echo "Resource upload for {$promise->getFilePath()} failed with {$exception->getMessage()}";
    }
);
```

The polling for promise status happens automatically every 0.5 seconds.

As fetching for the promise status happens asynchronously and the order of processed promises is not cretain there are callbacks for your implementation to react on successful and failed statuses. Another perk is that further processing is not blocked until all promisses are resolved.

#### Deleting a resource

Docs: [https://developers.transifex.com/reference/delete_resources-resource-id](https://developers.transifex.com/reference/delete_resources-resource-id)

```php
$response = $resources->delete('resource-a');
```

### Translations

Translations translate the resource content to another languages.

Get the Translations connector for the bellow examples like so:
```php
use Autoborna\Transifex\Connector\Translations;

$translations = $transifex->getConnector(Translations::class);
\assert($translations instanceof Translations);
```

#### Upload a translation

[Uploading a translation](https://developers.transifex.com/reference/post_resource-translations-async-uploads) is also asynchronous. The worflow is similar to downloading translations or resources.

```php
$response = $translations->upload(self::RESOURCE_SLUG, 'cs', "something=\"Něco\"\n");
$promise  = $transifex->getApiConnector()->createPromise($response);
$promises = new SplQueue();
$promise->setFilePath('/some/file.ini'); // In the real world it is useful to map your file to this promise. Useful for later processing.
$promises->enqueue($promise); // In the real world, there would be multiple promises.
usleep(500000); // Give Transifex a 1/2 second so we make 1 request instead of 2.
$transifex->getApiConnector()->fulfillPromises(
    $promises,
    function (ResponseInterface $response, Promise $promise) {
        echo "Translation for {$promise->getFilePath()} was uploaded successfully";
    },
    function (ResponseException $exception, Promise $promise) {
        echo "Translation upload for {$promise->getFilePath()} failed with {$exception->getMessage()}";
    }
);
```

#### Download a translation

This is the endpoint you need to get the work of the translators back to your app. Also an async operation. [Check the docs](https://developers.transifex.com/reference/post_resource-translations-async-downloads). It uses the same pattern as the others:

```php
$response = $translations->download(self::RESOURCE_SLUG, 'cs');
$promise  = $transifex->getApiConnector()->createPromise($response);
$promises = new SplQueue();
$promise->setFilePath('/some/file.ini'); // In the real world it is useful to map your file to this promise. Useful for later processing.
$promises->enqueue($promise); // In the real world, there would be multiple promises.

usleep(500000); // Give Transifex a 1/2 second so we make 1 request instead of 2.


// Assert that the translation content was downloaded successfully.
$transifex->getApiConnector()->fulfillPromises(
    $promises,
    function (ResponseInterface $response) use (&$successCounter, &$translationContent) {
        $translationContent = $response->getBody()->__toString();
        echo "Translation for {$promise->getFilePath()} was downloaded successfully. Here's the content:\n{$translationContent}";
    },
    function (ResponseException $exception) {
        echo "Translation download for {$promise->getFilePath()} failed with {$exception->getMessage()}";
    }
);
```

### Tests

There is one functional test that can run all the covered API requests against an existing Transifex project. It needs a configuration to connect to that project. Copy the `phpunit.xml.dist` file and rename it to `phpunit.xml`. This creates an override of the default PHPUNIT configuration. Fill in the organization, project and API token values. The test project must have the `cs` translation available.

To run the live test execute `composer test -- --testsuite=Functional`
To run the unit tests execute `composer test -- --testsuite=Unit`

Run the whole test suite: `composer test`
Run static analysis checks: `composer phpstan`
Run code style checks: `composer cs`
Fix code style issues: `composer fixcs`


CHANGELOG.markdown
CHANGELOG for the Transifex API Package
===============

See https://github.com/autoborna/Transifex-API/releases for newer releases.

* 3.0.0 (2019-11-19)

 * Raised minimum PHP version to 7.2
 * Add support for the Organizations API
 * Renamed `TransifexObject` to `ApiConnector`
 * Moved all connector classes to the `BabDev\Transifex\Connector` namespace
 * Made all classes final except the abstract `ApiConnector`
 * The `Transifex` class now requires a `FactoryInterface` for building API connectors as a constructor argument
 * The `ApiConnector` class now requires `RequestFactoryInterface` and `UriFactoryInterface` (PSR-17) implementations as constructor arguments
 * The `$client` constructor argument of the `ApiConnector` class is now a PSR-18 `ClientInterface` and is required
 * Removed `ApiConnector::getAuthData()`, a request's Authorization header will be added with the new `ApiConnector::createRequest()` method
 * `Transifex::setOption()` no longer has a return

* 2.1.1 (2018-03-10)

 * Code incorrectly referenced `base_url` as the option name versus `base_uri`

* 2.1.0 (2018-03-10)

 * [#7](https://github.com/BabDev/Transifex-API/pulls/7) Add the ?details fragment when retrieving language details
 * [#9](https://github.com/BabDev/Transifex-API/issues/9) The base_url option is not taken into account when issuing requests

* 2.0.0 (2016-12-21)

 * Dropped support for PHP 5.x, PHP 7.0 is the minimum requirement
 * Replace `joomla/http` package with `guzzlehttp/guzzle` for HTTP adapter
 * Add support for custom namespaces to `Transifex::get()`
 * Refactor API connector methods to return a full `Psr\Http\Message\ResponseInterface` object
 * Removed magic getter in `Transifex` and associated class member vars
 * Support a default option in `Transifex::getOption()`
 * `Translationstrings::getStrings()` now typehints the `$options` parameter
 * Rename `api.url` option to `base_uri`
 * Remove support for `ArrayAccess` objects as an options param, must pass an array
 * Removed `TransifexObject::fetchUrl()`, use the Guzzle API instead
 * Removed `TransifexObject::processResponse()`, use the Guzzle API instead

* 1.3.0 (2016-06-09)

 * Deprecated `Http` class and `joomla/http` package integration
 * Deprecated `TransifexObject::fetchUrl()`, 2.0 will use the Guzzle API to replace this functionality

* 1.2.0 (2015-07-13)

 * Add `Transifex::get()` to fetch API objects
 * Deprecated magic getter in `Transifex` and associated class member vars for storing objects

* 1.1.0 (2015-07-12)

 * Deprecated `TransifexObject::processResponse()`, 2.0 will return the full `Joomla\Http\Response` object instead of processing the response internally

* 1.0.0 (2014-10-20)

 * Initial stable release


src/ApiConnector.php
<?php declare(strict_types=1);

namespace Autoborna\Transifex;

use Autoborna\Transifex\Exception\MissingCredentialsException;
use Autoborna\Transifex\Exception\NotReadyException;
use Autoborna\Transifex\Exception\ResponseException;
use Psr\Http\Client\ClientInterface;
use Psr\Http\Message\RequestFactoryInterface;
use Psr\Http\Message\RequestInterface;
use Psr\Http\Message\ResponseInterface;
use Psr\Http\Message\StreamFactoryInterface;
use Psr\Http\Message\StreamInterface;
use Psr\Http\Message\UriFactoryInterface;
use Psr\Http\Message\UriInterface;
use SplQueue;

final class ApiConnector
{
    private ClientInterface $client;
    private RequestFactoryInterface $requestFactory;
    private StreamFactoryInterface $streamFactory;
    private UriFactoryInterface $uriFactory;
    private ConfigInterface $config;

    public function __construct(
        ClientInterface $client,
        RequestFactoryInterface $requestFactory,
        StreamFactoryInterface $streamFactory,
        UriFactoryInterface $uriFactory,
        ConfigInterface $config
    ) {
        $config->validate();

        $this->client         = $client;
        $this->requestFactory = $requestFactory;
        $this->streamFactory  = $streamFactory;
        $this->uriFactory     = $uriFactory;
        $this->config         = $config;
    }

    /**
     * Create a Request object for the given URI
     *
     * This method will also set the Authorization header for the request
     */
    public function createRequest(string $method, UriInterface $uri): RequestInterface
    {
        $request = $this->requestFactory->createRequest($method, $uri);
        $request = $request->withHeader('accept', 'application/vnd.api+json');
        $request = $request->withHeader('content-type', 'application/vnd.api+json');

        return $request->withHeader('authorization', $this->createAuthorizationHeader());
    }

    public function createUri(string $endpoint): UriInterface
    {
        if (\substr($endpoint, 0, 8) === 'https://') {
            $link = $endpoint;
        } else {
            $link = "{$this->config->getBaseUri()}/{$endpoint}";
        }

        return $this->uriFactory->createUri($link);
    }

    public function createStream(string $content): StreamInterface
    {
        return $this->streamFactory->createStream($content);
    }

    /**
     * Send a request to the API.
     *
     * @throws ResponseException if the response is not successful
     */
    public function sendRequest(RequestInterface $request): ResponseInterface
    {
        $response = $this->client->sendRequest($request);

        if ($response->getStatusCode() >= 400) {
            throw new ResponseException($request, $response);
        }

        return $response;
    }

    public function buildResourceString(string $resource): string
    {
        return "{$this->buildProjectString()}:r:{$resource}";
    }

    public function buildProjectString(): string
    {
        return "o:{$this->config->getOrganization()}:p:{$this->config->getProject()}";
    }

    /**
     * Helper method to create a Promise for further processing.
     */
    public function createPromise(ResponseInterface $response): Promise
    {
        $body = \json_decode($response->getBody()->__toString(), true);

        return new Promise($body['data']['id'], $body['data']['links']['self']);
    }

    /**
     * @param SplQueue<Promise>                          $promises
     * @param callable(ResponseInterface, Promise): void $onSuccess
     * @param callable(ResponseException, Promise): void $onError
     */
    public function fulfillPromises(SplQueue $promises, callable $onSuccess, callable $onError): void
    {
        $notReady = new SplQueue();

        while (!$promises->isEmpty()) {
            /** @var Promise $promise */
            $promise = $promises->dequeue();

            try {
                $onSuccess($this->getPromiseStatus($promise), $promise);
            } catch (NotReadyException $notReadyException) {
                $notReady->enqueue($promise);
            } catch (ResponseException $responseException) {
                $onError($responseException, $promise);
            }
        }

        if (!$notReady->isEmpty()) {
            \usleep(500000);
            $this->fulfillPromises($notReady, $onSuccess, $onError);
        }
    }

    public function getPromiseStatus(Promise $promise): ResponseInterface
    {
        $request  = $this->createRequest('GET', $this->createUri($promise->getCheckLink()));
        $response = $this->sendRequest($request);

        if ($response->getStatusCode() === 303) {
            $request  = $this->createRequest('GET', $this->createUri($response->getHeader('location')[0]));
            $response = $this->sendRequest($request);
        }

        $body   = $response->getBody()->__toString();
        $data   = \json_decode($body, true);
        $isJson = \json_last_error() === \JSON_ERROR_NONE;

        if ($isJson && isset($data['data']['attributes']['status'])) {
            if (\in_array($data['data']['attributes']['status'], ['pending', 'processing'])) {
                throw new NotReadyException($request, $response);
            }

            if (\in_array($data['data']['attributes']['status'], ['failed'])) {
                throw new ResponseException($request, $response);
            }
        }

        return $response;
    }

    /**
     * Creates the Authorization header for the request
     *
     * @throws MissingCredentialsException if credentials are not set
     */
    private function createAuthorizationHeader(): string
    {
        if (!$this->config->getApiToken()) {
            throw new MissingCredentialsException('Missing API token API authentication.');
        }

        return "Bearer {$this->config->getApiToken()}";
    }
}
